File: pipeline/generate_data.py
import pandas as pd
import numpy as np
from datetime import datetime, timedelta
import random
import os

# Configuration
START_DATE = datetime.now() - timedelta(days=365)
END_DATE = datetime.now()
DAYS = (END_DATE - START_DATE).days
STORES = [f"S{i:03d}" for i in range(1, 6)]
CHANNELS = ["Retail", "Ecommerce"]
SKUS = {
    "BEER_LAGER_6PK": {"name": "Classic Lager 6-Pack", "cat": "Beer", "price": 10.99, "lead_time": 7},
    "BEER_IPA_4PK": {"name": "Hoppy IPA 4-Pack", "cat": "Beer", "price": 12.99, "lead_time": 14},
    "BEER_STOUT_6PK": {"name": "Midnight Stout 6-Pack", "cat": "Beer", "price": 11.99, "lead_time": 10},
    "ALE_PALE_6PK": {"name": "Sunny Pale Ale 6-Pack", "cat": "Beer", "price": 10.49, "lead_time": 7},
    "UNKNOWN_SKU_999": {"name": "Discontinued Brew", "cat": "Legacy", "price": 5.00, "lead_time": 30}, # For testing DQ
}

np.random.seed(42)

def generate_sku_map():
    data = []
    for sku, info in SKUS.items():
        data.append({
            "sku": sku,
            "product_name": info["name"],
            "category": info["cat"],
            "pack_size": "6PK" if "6PK" in sku else "4PK",
            "active_flag": True if "UNKNOWN" not in sku else False
        })
    df = pd.DataFrame(data)
    df.to_csv("data/raw/sku_map.csv", index=False)
    print("Generated sku_map.csv")

def generate_pos_sales():
    data = []
    # Dates
    dates = [START_DATE + timedelta(days=i) for i in range(DAYS)]
    
    for date in dates:
        # Seasonality: higher in summer (months 6-8) and Dec (12)
        seasonality = 1.0
        if date.month in [6, 7, 8]: seasonality = 1.3
        if date.month == 12: seasonality = 1.2
        if date.weekday() >= 5: seasonality *= 1.5 # Weekends higher
        
        for store in STORES:
            for sku, info in SKUS.items():
                if "UNKNOWN" in sku and random.random() > 0.05: continue # Rare unknown sku
                
                # Base demand
                lambda_val = 5 * seasonality
                units = np.random.poisson(lambda_val)
                
                # Messy Data: Negatives
                if random.random() < 0.01: units = -1 * units
                
                # Messy Data: Promo
                promo = random.choice([True, False]) if random.random() < 0.1 else False
                if promo and units > 0: units =int(units * 1.5)
                
                if units != 0:
                    data.append({
                        "date": date.strftime("%Y-%m-%d"),
                        "store_id": store,
                        "sku": sku,
                        "units_sold": units,
                        "unit_price": info["price"] if not promo else info["price"] * 0.8,
                        "promo_flag": promo
                    })
    
    # Introduce duplicates
    data.extend(random.sample(data, 20))
    
    df = pd.DataFrame(data)
    df.to_csv("data/raw/pos_sales.csv", index=False)
    print(f"Generated pos_sales.csv with {len(df)} rows")

def generate_ecommerce_sales():
    data = []
    dates = [START_DATE + timedelta(days=i) for i in range(DAYS)]
    
    for date in dates:
        for sku, info in SKUS.items():
            if "UNKNOWN" in sku: continue
            
            # Ecommerce is more volatile
            if random.random() < 0.3: continue # Not sold every day
            
            units = np.random.randint(1, 20)
            discount = 0.0
            if random.random() < 0.05: discount = info["price"] * 0.1
            
            data.append({
                "date": date.strftime("%Y-%m-%d"),
                "sku": sku,
                "units_sold": units,
                "unit_price": info["price"],
                "discount": discount
            })
            
    df = pd.DataFrame(data)
    df.to_csv("data/raw/ecommerce_sales.csv", index=False)
    print(f"Generated ecommerce_sales.csv with {len(df)} rows")

def generate_inventory():
    data = []
    # Only need current inventory really, but let's generate a snapshot
    for sku, info in SKUS.items():
        data.append({
            "date": END_DATE.strftime("%Y-%m-%d"),
            "sku": sku,
            "on_hand": np.random.randint(0, 200),
            "on_order": np.random.randint(0, 100) if random.random() > 0.5 else 0,
            "lead_time_days": info["lead_time"]
        })
    df = pd.DataFrame(data)
    df.to_csv("data/raw/inventory.csv", index=False)
    print("Generated inventory.csv")

if __name__ == "__main__":
    os.makedirs("data/raw", exist_ok=True)
    generate_sku_map()
    generate_pos_sales()
    generate_ecommerce_sales()
    generate_inventory()

File: pipeline/generate_data.ts

import fs from 'fs';
import path from 'path';
import { addDays, subDays, startOfWeek, format } from 'date-fns';

const RAW_DIR = path.join(__dirname, '..', 'data', 'raw');
if (!fs.existsSync(RAW_DIR)) {
  fs.mkdirSync(RAW_DIR, { recursive: true });
}

// Config
const DAYS_HISTORY = 365;
const START_DATE = subDays(new Date(), DAYS_HISTORY);
const STORES = ['S001', 'S002', 'S003', 'S004', 'S005'];
const SKUS: Record<string, {name: string, cat: string, price: number, lead: number}> = {
    "BEER_LAGER_6PK": {name: "Classic Lager 6-Pack", cat: "Beer", price: 10.99, lead: 7},
    "BEER_IPA_4PK": {name: "Hoppy IPA 4-Pack", cat: "Beer", price: 12.99, lead: 14},
    "BEER_STOUT_6PK": {name: "Midnight Stout 6-Pack", cat: "Beer", price: 11.99, lead: 10},
    "ALE_PALE_6PK": {name: "Sunny Pale Ale 6-Pack", cat: "Beer", price: 10.49, lead: 7},
    "UNKNOWN_SKU_999": {name: "Discontinued Brew", cat: "Legacy", price: 5.00, lead: 30},
};

function writeCsv(filename: string, headers: string[], rows: any[]) {
    const headerLine = headers.join(',');
    const lines = rows.map(r => headers.map(h => r[h]).join(','));
    fs.writeFileSync(path.join(RAW_DIR, filename), [headerLine, ...lines].join('\n'));
    console.log(`Generated ${filename} with ${rows.length} rows`);
}

// 1. SKU Map
const skuRows = Object.entries(SKUS).map(([sku, info]) => ({
    sku,
    product_name: info.name,
    category: info.cat,
    pack_size: sku.includes('6PK') ? '6PK' : '4PK',
    active_flag: !sku.includes('UNKNOWN')
}));
writeCsv('sku_map.csv', ['sku', 'product_name', 'category', 'pack_size', 'active_flag'], skuRows);

// 2. POS Sales
const posRows: any[] = [];
for (let i = 0; i < DAYS_HISTORY; i++) {
    const date = addDays(START_DATE, i);
    const dateStr = format(date, 'yyyy-MM-dd');
    const isWeekend = date.getDay() === 0 || date.getDay() === 6;
    const isSummer = date.getMonth() >= 5 && date.getMonth() <= 7;
    const isDec = date.getMonth() === 11;

    let seasonality = 1.0;
    if (isSummer) seasonality = 1.3;
    if (isDec) seasonality = 1.2;
    if (isWeekend) seasonality *= 1.5;

    STORES.forEach(store => {
        Object.keys(SKUS).forEach(sku => {
            if (sku.includes('UNKNOWN') && Math.random() > 0.05) return;
            
            let lambda = 5 * seasonality;
            // Simple Poisson-ish approximation
            let units = Math.floor(lambda + (Math.random() - 0.5) * lambda); 
            if (units < 0) units = 0;

            // Messy: Negative
            if (Math.random() < 0.01) units = -units;

            // Messy: Promo
            let promo = false;
            let price = SKUS[sku].price;
            if (Math.random() < 0.1) {
                promo = true;
                if (units > 0) units = Math.floor(units * 1.5);
                price *= 0.8;
            }

            if (units !== 0) {
                posRows.push({
                    date: dateStr,
                    store_id: store,
                    sku,
                    units_sold: units,
                    unit_price: price.toFixed(2),
                    promo_flag: promo
                });
            }
        });
    });
}
// Duplicates
posRows.push(...posRows.slice(0, 20));
writeCsv('pos_sales.csv', ['date', 'store_id', 'sku', 'units_sold', 'unit_price', 'promo_flag'], posRows);

// 3. Ecommerce Sales
const ecomRows: any[] = [];
for (let i = 0; i < DAYS_HISTORY; i++) {
    const date = addDays(START_DATE, i);
    const dateStr = format(date, 'yyyy-MM-dd');

    Object.keys(SKUS).forEach(sku => {
        if (sku.includes('UNKNOWN')) return;
        if (Math.random() < 0.3) return; // Sparse

        const units = Math.floor(Math.random() * 20) + 1;
        const discount = Math.random() < 0.05 ? (SKUS[sku].price * 0.1).toFixed(2) : 0;
        
        ecomRows.push({
            date: dateStr,
            sku,
            units_sold: units,
            unit_price: SKUS[sku].price,
            discount
        });
    });
}
writeCsv('ecommerce_sales.csv', ['date', 'sku', 'units_sold', 'unit_price', 'discount'], ecomRows);

// 4. Inventory
const invRows = Object.entries(SKUS).map(([sku, info]) => ({
    date: format(new Date(), 'yyyy-MM-dd'),
    sku,
    on_hand: Math.floor(Math.random() * 200),
    on_order: Math.random() > 0.5 ? Math.floor(Math.random() * 100) : 0,
    lead_time_days: info.lead
}));
writeCsv('inventory.csv', ['date', 'sku', 'on_hand', 'on_order', 'lead_time_days'], invRows);
